File: backend/src/main/java/com/avocarbon/platform/service/ProjectService.java
package com.avocarbon.platform.service;

import com.avocarbon.platform.dto.ProjectRequest;
import com.avocarbon.platform.dto.ProjectResponse;

import java.util.List;

/**
 * Service interface for project operations.
 *
 * Defines the business logic contract for managing projects.
 * Separates the service layer from implementation details.
 *
 * @since 1.0.0
 */
public interface ProjectService {

    /**
     * Create a new project.
     *
     * @param request the project creation request
     * @return the created project response
     */
    ProjectResponse createProject(ProjectRequest request);

    /**
     * Retrieve all projects.
     *
     * @return list of all projects
     */
    List<ProjectResponse> getAllProjects();

    /**
     * Retrieve a project by ID.
     *
     * @param id the project ID
     * @return the project response
     * @throws com.avocarbon.platform.exception.ResourceNotFoundException if project not found
     */
    ProjectResponse getProjectById(Long id);

    /**
     * Update an existing project.
     *
     * @param id the project ID
     * @param request the project update request
     * @return the updated project response
     * @throws com.avocarbon.platform.exception.ResourceNotFoundException if project not found
     */
    ProjectResponse updateProject(Long id, ProjectRequest request);

    /**
     * Delete a project by ID.
     *
     * @param id the project ID
     * @throws com.avocarbon.platform.exception.ResourceNotFoundException if project not found
     */
    void deleteProject(Long id);

}

File: frontend/src/app/core/services/project.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ProjectRequest, ProjectResponse } from '../models/project.model';

@Injectable({
  providedIn: 'root'
})
export class ProjectService {
  private http = inject(HttpClient);
  private apiUrl = 'http://localhost:8080/api/projects';

  getAllProjects(): Observable<ProjectResponse[]> {
    return this.http.get<ProjectResponse[]>(this.apiUrl);
  }

  getProjectById(id: number): Observable<ProjectResponse> {
    return this.http.get<ProjectResponse>(`${this.apiUrl}/${id}`);
  }

  createProject(request: ProjectRequest): Observable<ProjectResponse> {
    return this.http.post<ProjectResponse>(this.apiUrl, request);
  }

  updateProject(id: number, request: ProjectRequest): Observable<ProjectResponse> {
    return this.http.put<ProjectResponse>(`${this.apiUrl}/${id}`, request);
  }

  deleteProject(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
